### DISCUSSION TITLE
OSError: [WinError 6] The handle is invalid (sys.stdout.write)

### DISCUSSION THREAD

**Discussion by pllim** [pytest-dev/pytest#10976] [Q&A]
I don't run the test suite locally on Windows often and the Windows job in Actions does not see this error. While Actions use tox, locally I use conda env without tox to run the tests. I am seeing https://github.com/astropy/astropy/issues/14791 that I do not know how to debug, so any idea would be greatly appreciated.

You could reproduce this error either from our RC or from dev version built from source. Example RC install:

```
pip install "astropy[test]==5.3rc1"
```

Then you can run a subset of the test suite that would give you these errors on Windows:

```
python -c "import astropy; astropy.test(package='io.votable')"
```

If I drop into an interactive Python session, I can still see this error when I run pytest like this:

```python
>>> import astropy
>>> astropy.test(package="io.votable")
```

However, in the same session, there is no error if I run one of the failing test case on its own without pytest:

```python
>>> from astropy.io.votable.tests.table_test import test_validate_output_invalid
>>> test_validate_output_invalid()
```

I could not run pytest from CLI like this because it is throwing `ValueError: Plugin already registered: ...\astropy\conftest.py`, which seems like an unrelated problem that I don't have time to dig into right now. But if you don't get the `ValueError` and you are able to build `astropy` from source, you can also do this locally from the source checkout directory:

```
pytest astropy/io/votable/tests/table_test.py
```

All the `astropy.test()` calls go through our test runner here:

https://github.com/astropy/astropy/blob/main/astropy/tests/runner.py

https://github.com/astropy/astropy/blob/eec06978bcf37dba7e998f198003a3b6cbfcfb94/astropy/__init__.py#L137

**Comment by portyu9** [comment]
Hi @pllim 

This looks like a **stale `sys.stdout` reference**, rather than a tox/conda difference.

The important detail is that the affected Astropy code uses `sys.stdout` as a function default. A default such as:

```python
def validate(..., output=sys.stdout, ...):
```

is evaluated when the function is defined. Pytest's output capture replaces/redirections stdout while tests run, so that saved object can later refer to a stream whose underlying Windows handle is no longer valid.

That also explains why calling the function directly outside pytest does not reproduce the problem.

The application-side fix is to avoid binding `sys.stdout` in the function signature:

```python
def validate(..., output=None, ...):
    if output is None:
        output = sys.stdout
```

As a diagnostic, running with:

```text
pytest -s
```

or:

```text
pytest --capture=no
```

should make a capture-related failure disappear, but disabling capture is only a workaround. The preferable fix is to resolve `sys.stdout` at call time as above.

So I would fix this in Astropy rather than trying to make the local environment match the tox environment.

Please mark this answer as accepted if it helped, thank you.
